0022-circuit-block.md

```diff
@@ -0,0 +1,300 @@
+
+The actual implementation of the object looks something like
+
+```python
+import typing
+from qiskit.circuit.controlflow import ControlFlowOp
+
+@typing.Final
+class Box(ControlFlowOp):
+    def __init__(self, circuit, annotations):
+        super().__init__("box", circuit.num_qubits, circuit.num_clbits, [circuit])
+        self.annotations = annotations
+
+    @property
+    def blocks(self):
+        return (self.params[0],)
+
+    def replace_blocks(self, blocks):
+        assert len(blocks) == 1
+        return Box(blocks[0], self.annotations.copy())
+```
+
+See below for more considerations about the ``annotations`` parameter.
+
+This then implies that Qiskit should implement the control-flow builder interfaces for `box`.
+Again, using some dummy "annotations" (which are not specified by this RFC, and are given for illustrative purposes only):
+```python
+from qiskit.circuit import QuantumCircuit
+
+# Hypothetical core-Qiskit box-level annotation.
+from qiskit.transpiler.contexts import ScheduleStrategy
+# Hypothetical annotations for IBM-specific primitives.
+from qiskit_ibm_runtime.contexts import TwirlingStrategy
+
+qc = QuantumCircuit(5, 5)
+with qc.box([ScheduleStrategy.ALAP, TwirlingStrategy(...)]):
+    qc.h(0)
+    qc.cx(0, 1)
+    qc.x(2)
+with qc.box([ScheduleStrategy.ASAP, TwirlingStrategy(...)]):
+    qc.h(2)
+    qc.cx(2, 3)
+    qc.x(4)
+qc.measure(qc.qubits, qc.clbits)
+```
+
+For `box` in particular (but also for other control-flow operations), it can be useful to mark certain qubits and other circuit resources as being explicitly "used" by a block, even if the actual operation is a no-op.
+For example, twirling and noise learning may want to consider certain qubits along with a complete block, even if that qubit is undergoing the logical identity for the span.
+This can be worked around by inserting explicit `id` gates into the box, or some other form of no-op operation, but these instructions can sometimes be treated by hardware as something other than a no-op, which might cause performance problems if the compiler fails to optimise them out or understand their intention.
+For example, IBM hardware has, in the past, treated the `id` instruction as a single-cycle delay on a single qubit, which is not precisely equal to a no-op for the purposes of scheduling, especially if the rest of the block has a natural duration of zero cycles (such as if it comprises only virtual phase update `rz` gates).
+We also do not currently have the concept of a no-op on classical data.
+To make the user intent clearer here, we propose adding a `QuantumCircuit.noop` to make the intent explicit:
+
+> [!NOTE]
+> *Bikeshedding potential*: do you prefer `QuantumCircuit.use` or `QuantumCircuit.noop` for this?
+> `use` is perhaps clearer for more users, but its name only makes sense in a control-flow builder context, whereas `noop` makes sense to appear in the root scope of a `QuantumCircuit`.
+
```

Comment:
We should probably remove this before opening the PR

```suggestion
```

Reply:
i vote to open the rfc with `noop`